src/work.cl

```diff
@@ -1,349 +1,104 @@
-  v[13] = S->t[1] ^ blake2b_IV[5];
-  v[14] = S->f[0] ^ blake2b_IV[6];
-  v[15] = S->f[1] ^ blake2b_IV[7];
-#define G(r,i,a,b,c,d) \
-  do { \
-	a = a + b + m[blake2b_sigma[r][2*i+0]]; \
-	d = rotr64(d ^ a, 32); \
-	c = c + d; \
-	b = rotr64(b ^ c, 24); \
-	a = a + b + m[blake2b_sigma[r][2*i+1]]; \
-	d = rotr64(d ^ a, 16); \
-	c = c + d; \
-	b = rotr64(b ^ c, 63); \
-  } while(0)
-#define ROUND(r)  \
-  do { \
-	G(r,0,v[ 0],v[ 4],v[ 8],v[12]); \
-	G(r,1,v[ 1],v[ 5],v[ 9],v[13]); \
-	G(r,2,v[ 2],v[ 6],v[10],v[14]); \
-	G(r,3,v[ 3],v[ 7],v[11],v[15]); \
-	G(r,4,v[ 0],v[ 5],v[10],v[15]); \
-	G(r,5,v[ 1],v[ 6],v[11],v[12]); \
-	G(r,6,v[ 2],v[ 7],v[ 8],v[13]); \
-	G(r,7,v[ 3],v[ 4],v[ 9],v[14]); \
-  } while(0)
-  ROUND( 0 );
-  ROUND( 1 );
-  ROUND( 2 );
-  ROUND( 3 );
-  ROUND( 4 );
-  ROUND( 5 );
-  ROUND( 6 );
-  ROUND( 7 );
-  ROUND( 8 );
-  ROUND( 9 );
-  ROUND( 10 );
-  ROUND( 11 );
-
-  for( i = 0; i < 8; ++i )
-	S->h[i] = S->h[i] ^ v[i] ^ v[i + 8];
-
-#undef G
+#define G32(m0, m1, m2, m3, vva, vb1, vb2, vvc, vd1, vd2) \
+    do {                                                  \
+        vva += (ulong2)(vb1 + m0, vb2 + m2);              \
+        vd1 = rotr64(vd1 ^ vva.s0, 32);                   \
+        vd2 = rotr64(vd2 ^ vva.s1, 32);                   \
+        vvc += (ulong2)(vd1, vd2);                        \
+        vb1 = rotr64(vb1 ^ vvc.s0, 24);                   \
+        vb2 = rotr64(vb2 ^ vvc.s1, 24);                   \
+        vva += (ulong2)(vb1 + m1, vb2 + m3);              \
+        vd1 = rotr64(vd1 ^ vva.s0, 16);                   \
+        vd2 = rotr64(vd2 ^ vva.s1, 16);                   \
+        vvc += (ulong2)(vd1, vd2);                        \
+        vb1 = rotr64(vb1 ^ vvc.s0, 63);                   \
+        vb2 = rotr64(vb2 ^ vvc.s1, 63);                   \
+    } while (0)
+
+#define G2v(m0, m1, m2, m3, a, b, c, d)                                   \
+    G32(m0, m1, m2, m3, vv[a / 2], vv[b / 2].s0, vv[b / 2].s1, vv[c / 2], \
+        vv[d / 2].s0, vv[d / 2].s1)
+
+#define G2v_split(m0, m1, m2, m3, a, vb1, vb2, c, vd1, vd2) \
+    G32(m0, m1, m2, m3, vv[a / 2], vb1, vb2, vv[c / 2], vd1, vd2)
+
+#define ROUND(m0, m1, m2, m3, m4, m5, m6, m7, m8, m9, m10, m11, m12, m13, m14, \
+              m15)                                                             \
+    do {                                                                       \
+        G2v(m0, m1, m2, m3, 0, 4, 8, 12);                                      \
+        G2v(m4, m5, m6, m7, 2, 6, 10, 14);                                     \
+        G2v_split(m8, m9, m10, m11, 0, vv[5 / 2].s1, vv[6 / 2].s0, 10,         \
+                  vv[15 / 2].s1, vv[12 / 2].s0);                               \
+        G2v_split(m12, m13, m14, m15, 2, vv[7 / 2].s1, vv[4 / 2].s0, 8,        \
+                  vv[13 / 2].s1, vv[14 / 2].s0);                               \
+    } while (0)
+
+static inline ulong blake2b(ulong const nonce, ulong *const h)
```

Comment:
```suggestion
static inline ulong blake2b(ulong const nonce, __constant ulong *h)
```

@jserv broke for Nvidia runtime. Tried with this change on both AMD and Nvidia, seems to finall be ok with this one.

Feel free to not force-push since we'll be squash committing anyway.